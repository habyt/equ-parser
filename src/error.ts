
export class EQUError extends Error {
    constructor(msg?: string) {
        super(msg)
    }
}

export class EQULexingError extends EQUError {
    constructor(msg?: string) {
        super(msg)
    }
}

export class EQUParsingError extends EQUError {
    constructor(msg?: string) {
        super(msg)
    }
}

