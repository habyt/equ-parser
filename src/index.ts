export * from "./equ"

export * from "./error"

export { Token, Item } from "./lexer"
export {
    FilterOperator,
    FilterOperand,
    ParseItem,
    ExpressionOperand,
    ExpressionOperator,
    ExpressionItem,
} from "./parser"
