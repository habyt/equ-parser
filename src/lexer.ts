import assert from "assert"
export type Token =
    | "eof"
    | "error"
    | "path"
    | "or"
    | "and"
    | "filtersStart"
    | "filtersEnd"
    | "bundleStart"
    | "bundleEnd"
    | "string"
    | "number"
    | "filterGt"
    | "filterLt"
    | "filterGte"
    | "filterLte"
    | "filterEq"
    | "filterCt"

const filters: { [k: string]: Token } = {
    eq: "filterEq",
    lt: "filterLt",
    gt: "filterGt",
    lte: "filterLte",
    gte: "filterGte",
    ct: "filterCt"
}

const tokenBundleStart = "("
const tokenBundleEnd = ")"

const tokenOr = "|"
const tokenAnd = ","

const tokenFiltersStart = "["
const tokenFiltersEnd = "]"

const tokenFilterIs = ":"

const tokenValidPath =
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._"
const tokenValidOperator = "abcdefghijklmnopqrstuvwxyz"
const tokenValidNumberStart = "+-.0123456789"

type StateFn = ((lexer: Lexer) => StateFn) | undefined

export class EQUParsingError extends Error {
    constructor(msg?: string) {
        super(msg)
    }
}

function lexError(msg?: string): never {
    throw new EQUParsingError(msg)
}

class Lexer {
    input: string
    pos: number = 0
    items: Array<Item> = []
    start: number = 0
    width: number = 0

    constructor(str: string) {
        this.input = str
    }

    emit(token: Token) {
        let value = this.input.substring(this.start, this.pos)

        if (token === "string") {
            value = value.replace("\\\\", "\\")
            value = value.replace('\\"', '"')
        }

        this.items.push({ str: value, type: token })
        this.start = this.pos
    }

    next(): string {
        const next = this.input.substr(this.pos, 1)
        this.width = next.length
        this.pos += this.width
        return next
    }

    accept(valid: string): boolean {
        if (valid.indexOf(this.next()) >= 0) {
            return true
        }

        this.backup()
        return false
    }

    backup() {
        this.pos -= this.width
    }

    peek(): string {
        const next = this.next()
        this.backup()
        return next
    }

    acceptRun(valid: string) {
        while (true) {
            const next = this.next()
            if (next === "") {
                break
            }

            if (valid.indexOf(next) === -1) {
                break
            }
        }

        this.backup()
    }

    acceptRunUntil(...stops: Array<string>) {
        while (true) {
            const next = this.next()
            if (next === "") {
                this.backup()
                return false
            }

            if (stops.indexOf(next) >= 0) {
                this.backup()
                return true
            }
        }
    }

    ignore() {
        this.start = this.pos
    }

    rest(): string {
        return this.input.substring(this.start)
    }

    current(): string {
        return this.input.substring(this.start, this.pos)
    }
}

export type Item = {
    str: string
    type: Token
}

function lexFilters(l: Lexer): StateFn {
    if (l.peek() === tokenBundleStart) {
        return lexToken(tokenBundleStart, "bundleStart", lexFilters)
    }

    l.acceptRun(tokenValidPath)
    if (l.pos <= l.start) {
        throw new EQUParsingError("expected path, but got " + l.rest())
    }

    l.emit("path")

    if (l.peek() !== tokenFiltersStart) {
        throw new EQUParsingError("expected '[', but got " + l.rest())
    }

    return lexToken(tokenFiltersStart, "filtersStart", lexInsideFilters)
}

function lexInsideFilters(l: Lexer): StateFn {
    if (l.peek() === tokenBundleStart) {
        return lexToken(tokenBundleStart, "bundleStart", lexInsideFilters)
    }

    l.acceptRun(tokenValidOperator)
    const nextFilter = l.current()
    const filterType =
        filters[nextFilter] ??
        lexError("expected filter type, got " + nextFilter)
    l.emit(filterType)

    if (l.next() !== tokenFilterIs) {
        lexError("expected " + tokenFilterIs + " but got " + l.current())
    }

    l.ignore()

    return lexFilterValue
}

function lexFilterValue(l: Lexer): StateFn {
    const next = l.peek()

    if (tokenValidNumberStart.indexOf(next) >= 0) {
        return lexNumberValue
    }

    if (next === '"') {
        l.next()
        l.ignore()
        return lexStringValue
    }

    lexError("expected filter value but got " + l.rest())
}

function lexStringValue(l: Lexer): StateFn {
    while (true) {
        const found = l.acceptRunUntil("\\", '"')
        if (!found) {
            lexError("expected string to end, but got " + l.rest())
        }

        const next = l.peek()
        if (next === '"') {
            l.emit("string")

            l.next()
            l.ignore()
            return lexAfterFilterValue
        }

        // if we're here next must be a \ because of acceptRunUntil
        assert(next === "\\")
        l.next()

        const escapedSymbol = l.peek()

        if (escapedSymbol === "\\" || escapedSymbol === '"') {
            l.next()
        } else {
            lexError("invalid escape sequence: \\" + escapedSymbol)
        }
    }
}

function lexNumberValue(l: Lexer): StateFn {
    l.accept("+-")

    let decimalOccured = l.accept(".")

    l.acceptRun("0123456789")
    if (l.accept(".")) {
        if (decimalOccured) {
            lexError("invalid number: " + l.current())
        }

        l.acceptRun("0123456789")
    }

    l.emit("number")
    return lexAfterFilterValue
}

function lexAfterFilterValue(l: Lexer): StateFn {
    const next = l.next()
    switch (next) {
        case tokenOr: {
            l.emit("or")
            break
        }
        case tokenAnd: {
            l.emit("and")
            break
        }
        case tokenFiltersEnd: {
            l.backup()
            return lexToken(tokenFiltersEnd, "filtersEnd", lexAfterFilter)
        }
        case tokenBundleEnd: {
            l.backup()
            return lexToken(tokenBundleEnd, "bundleEnd", lexAfterFilterValue)
        }
    }

    return lexInsideFilters
}

function lexAfterFilter(l: Lexer): StateFn {
    const next = l.next()

    if (next === "") {
        l.emit("eof")
        return undefined
    }

    switch (next) {
        case tokenOr: {
            l.emit("or")
            return lexFilters
        }
        case tokenAnd: {
            l.emit("and")
            return lexFilters
        }
        case tokenBundleEnd: {
            l.backup()
            return lexToken(tokenBundleEnd, "bundleEnd", lexAfterFilter)
        }
    }

    lexError("invalid combinator: " + next)
}

function lexToken(tokenValue: string, token: Token, next: StateFn): StateFn {
    return (lexer: Lexer) => {
        lexer.pos += tokenValue.length
        lexer.emit(token)
        return next
    }
}

export type LexerOptions = {
    debug?: boolean
}

export function lex(str: string): Array<Item> {
    const lexer = new Lexer(str)

    let state: StateFn = lexFilters
    while (state !== undefined) {
        state = state(lexer)
    }

    return lexer.items
}
