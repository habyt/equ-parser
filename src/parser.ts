import { Item, EQUParsingError } from "./lexer"

function expressionPrecedence(
    a: ExpressionOperator,
    b: ExpressionOperator
): boolean {
    const list = ["or", "and", "bundleStart", "bundleEnd"]

    return list.indexOf(a.operator) > list.indexOf(b.operator)
}

function filterPrecedence(a: FilterOperator, b: FilterOperator): boolean {
    const list = ["or", "and", "bundleStart", "bundleEnd"]

    return list.indexOf(a.operator) > list.indexOf(b.operator)
}

function parseError(msg?: any): never {
    if (msg === undefined) {
        throw new EQUParsingError()
    }

    if (typeof msg === "string") {
        throw new EQUParsingError(msg)
    }

    throw new EQUParsingError("unimplemented " + JSON.stringify(msg))
}

type StateFn = ((ctx: ParserContext) => StateFn) | undefined

type FilterOperatorType = "and" | "or" | "bundleStart" | "bundleEnd"

type ExpressionType = "eq" | "gt" | "gte" | "lt" | "lte" | "ct"
type ExpressionOperatorType = "and" | "or" | "bundleStart" | "bundleEnd"

type ValueType = "number" | "string"

export type FilterOperator = {
    type: "operator"
    operator: FilterOperatorType
}

export type FilterOperand = {
    type: "operand"
    path: string
    expressions: Array<ExpressionItem>
}

export type ParseItem = FilterOperand | FilterOperator

export type ExpressionOperand = {
    type: "expressionOperand"
    expressionType: ExpressionType
    valueType: ValueType
    value: string | number
}

export type ExpressionOperator = {
    type: "expressionOperator"
    operator: ExpressionOperatorType
}

export type ExpressionItem = ExpressionOperator | ExpressionOperand

class ParserContext {
    items: Array<Item>
    parsedItems: Array<ParseItem> = []

    filterOperatorStack: Array<FilterOperator> = []
    expressionOperatorStack: Array<ExpressionOperator> = []

    constructor(items: Array<Item>) {
        this.items = items
    }

    next(): Item {
        return this.items.shift() ?? parseError("expected to have more items")
    }

    pushFilterItem(item: ParseItem) {
        this.parsedItems.push(item)
    }

    pushExpressionItem(item: ExpressionItem) {
        const current =
            this.currentFilterOperand() ??
            parseError("expected to have a filter operand")

        current.expressions.push(item)
    }

    popFilterOperator(): FilterOperator | undefined {
        return this.filterOperatorStack.pop()
    }

    popExpressionOperator(): ExpressionOperator | undefined {
        return this.expressionOperatorStack.pop()
    }

    topExpressionOperator(): ExpressionOperator | undefined {
        if (this.expressionOperatorStack.length === 0) {
            return undefined
        }

        return this.expressionOperatorStack[
            this.expressionOperatorStack.length - 1
        ]
    }

    topFilterOperator(): FilterOperator | undefined {
        if (this.filterOperatorStack.length === 0) {
            return undefined
        }

        return this.filterOperatorStack[this.filterOperatorStack.length - 1]
    }

    currentFilterOperand(): FilterOperand | undefined {
        for (let i = this.parsedItems.length - 1; i >= 0; i--) {
            const item = this.parsedItems[i]
            if (item.type === "operand") {
                return item
            }
        }

        return undefined
    }
}

function parsePath(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "bundleStart") {
        return parseFilterBundleStart
    }

    if (next.type === "path") {
        ctx.pushFilterItem({
            type: "operand",
            path: next.str,
            expressions: []
        })

        return parseFiltersStart
    }

    parseError(next)
}

function parseFilterBundleStart(ctx: ParserContext): StateFn {
    ctx.filterOperatorStack.push({
        type: "operator",
        operator: "bundleStart"
    })

    return parsePath
}

function parseFiltersStart(ctx: ParserContext): StateFn {
    const next = ctx.next()
    if (next.type === "filtersStart") {
        return parseExpression
    }

    parseError(next)
}

function parseExpression(ctx: ParserContext): StateFn {
    const expressionType = ctx.next()

    if (expressionType.type === "bundleStart") {
        return parseExpressionBundleStart
    }

    let expressionOperator: ExpressionType = "eq"
    switch (expressionType.type) {
        case "filterCt": {
            expressionOperator = "ct"
            break
        }
        case "filterEq": {
            expressionOperator = "eq"
            break
        }
        case "filterGt": {
            expressionOperator = "gt"
            break
        }
        case "filterGte": {
            expressionOperator = "gte"
            break
        }
        case "filterLt": {
            expressionOperator = "lt"
            break
        }
        case "filterLte": {
            expressionOperator = "lte"
            break
        }
        default:
            parseError(expressionType)
    }

    const value = ctx.next()
    let valueType: "string" | "number" = "string"
    switch (value.type) {
        case "string": {
            valueType = "string"
            break
        }
        case "number": {
            valueType = "number"
            break
        }
        default:
            parseError(value)
    }

    let parsedValue: number | string = value.str
    if (value.type === "number") {
        parsedValue = Number(value.str)
    }

    ctx.pushExpressionItem({
        type: "expressionOperand",
        expressionType: expressionOperator,
        valueType,
        value: parsedValue
    })

    return parseAfterExpression
}

function parseExpressionBundleStart(ctx: ParserContext): StateFn {
    ctx.expressionOperatorStack.push({
        type: "expressionOperator",
        operator: "bundleStart"
    })

    return parseExpression
}

function parseAfterExpression(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "filtersEnd") {
        const currentFilterOperand = ctx.currentFilterOperand() ?? parseError()
        currentFilterOperand.expressions.push(
            ...ctx.expressionOperatorStack.reverse()
        )
        ctx.expressionOperatorStack = []

        return parseAfterFilter
    }

    if (next.type === "or" || next.type === "and") {
        const expressionOperator: ExpressionOperator = {
            type: "expressionOperator",
            operator: next.type
        }

        let currentTopExpression = ctx.topExpressionOperator()
        while (
            currentTopExpression !== undefined &&
            expressionPrecedence(expressionOperator, currentTopExpression)
        ) {
            const top = ctx.popExpressionOperator() ?? parseError()
            ctx.pushExpressionItem(top)
            currentTopExpression = ctx.topExpressionOperator()
        }

        ctx.expressionOperatorStack.push(expressionOperator)

        return parseExpression
    }

    if (next.type === "bundleEnd") {
        let op = ctx.popExpressionOperator()
        while (op !== undefined) {
            if (op.operator === "bundleStart") {
                break
            }

            ctx.pushExpressionItem(op)
            op = ctx.popExpressionOperator()
        }
        return parseAfterExpression
    }

    parseError(next)
}

function parseAfterFilter(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "eof") {
        ctx.parsedItems.push(...ctx.filterOperatorStack.reverse())
        ctx.filterOperatorStack = []

        return undefined
    }

    if (next.type === "and" || next.type === "or") {
        const filterOperator: FilterOperator = {
            type: "operator",
            operator: next.type
        }

        let currentTopFilter = ctx.topFilterOperator()
        while (
            currentTopFilter !== undefined &&
            filterPrecedence(filterOperator, currentTopFilter)
        ) {
            const top = ctx.popFilterOperator() ?? parseError()
            ctx.pushFilterItem(top)
            currentTopFilter = ctx.topFilterOperator()
        }

        ctx.filterOperatorStack.push(filterOperator)

        return parsePath
    }

    if (next.type === "bundleEnd") {
        let op = ctx.popFilterOperator()
        while (op !== undefined) {
            if (op.operator === "bundleStart") {
                break
            }

            ctx.pushFilterItem(op)
            op = ctx.popFilterOperator()
        }
        return parseAfterFilter
    }

    parseError(next)
}

export function parse(items: Array<Item>): Array<ParseItem> {
    const parser = new ParserContext(items)
    let state: StateFn = parsePath
    while (state !== undefined) {
        state = state(parser)
    }

    return parser.parsedItems
}
