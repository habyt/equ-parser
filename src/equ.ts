import { ParseItem } from "./parser"
import { lex } from "./lexer"
import { parse as parseLex } from "./parser"

export function parse(equ: string): Array<ParseItem> {
    const lexed = lex(equ)

    return parseLex(lexed)
}
