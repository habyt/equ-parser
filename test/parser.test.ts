import { parse } from "../src/parser"
import { Item } from "../src/lexer"

const example: Array<Item> = [
    { str: "name.first", type: "path" },
    { str: "[", type: "filtersStart" },
    { str: "eq", type: "filterEq" },
    { str: "foo", type: "string" },
    { str: "|", type: "or" },
    { str: "eq", type: "filterEq" },
    { str: "bar", type: "string" },
    { str: "]", type: "filtersEnd" },
    { str: "|", type: "or" },
    { str: "email", type: "path" },
    { str: "[", type: "filtersStart" },
    { str: "ct", type: "filterCt" },
    { str: "foo", type: "string" },
    { str: ",", type: "and" },
    { str: "ct", type: "filterCt" },
    { str: "bar", type: "string" },
    { str: "]", type: "filtersEnd" },
    { str: ",", type: "and" },
    { str: "age", type: "path" },
    { str: "[", type: "filtersStart" },
    { str: "(", type: "bundleStart" },
    { str: "gt", type: "filterGt" },
    { str: "4.5", type: "number" },
    { str: ",", type: "and" },
    { str: "lt", type: "filterLt" },
    { str: "-10", type: "number" },
    { str: ")", type: "bundleEnd" },
    { str: "|", type: "or" },
    { str: "eq", type: "filterEq" },
    { str: "15", type: "number" },
    { str: "]", type: "filtersEnd" },
    { str: "", type: "eof" }
]

describe("parser", () => {
    it("should parse the example", () => {
        const result = parse(example)

        expect(result).toMatchSnapshot()
    })

    it("should throw a parse error", () => {
        expect(() => parse([])).toThrowErrorMatchingSnapshot()
    })

    it("should throw a parse error on unimplemented items", () => {
        expect(() =>
            parse([{ str: "test", type: "error" }])
        ).toThrowErrorMatchingSnapshot()
    })

    it("should throw a parse error for parsing expressions without a filter", () => {
        expect(() =>
            parse([
                { str: "", type: "path" },
                { str: "test", type: "filtersStart" },
                { str: "test", type: "filtersEnd" }
            ])
        ).toThrowErrorMatchingSnapshot()
    })

    it("should parse bundles correctly", () => {
        const bundleExample: Array<Item> = [
            { str: "(", type: "bundleStart" },
            { str: "path", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "eq", type: "filterEq" },
            { str: "asd", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: ",", type: "and" },
            { str: "path2", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "eq", type: "filterEq" },
            { str: "bdf", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: ")", type: "bundleEnd" },
            { str: "", type: "eof" }
        ]

        const result = parse(bundleExample)
        expect(result).toMatchSnapshot()
    })

    it("should throw a parse error on missing filter start", () => {
        expect(() =>
            parse([
                { str: "test", type: "path" },
                { str: "bla", type: "error" }
            ])
        ).toThrowErrorMatchingSnapshot()
    })

    it("should parse gte statements", () => {
        const result = parse([
            { str: "path", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "gte", type: "filterGte" },
            { str: "value", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: "", type: "eof" }
        ])

        expect(result).toMatchSnapshot()
    })

    it("should parse nex statements", () => {
        const result = parse([
            { str: "path", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "nex", type: "filterNex" },
            { str: "value", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: "", type: "eof" }
        ])

        expect(result).toMatchSnapshot()
    })

    it("should parse neq statements", () => {
        const result = parse([
            { str: "path", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "neq", type: "filterNeq" },
            { str: "value", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: "", type: "eof" }
        ])

        expect(result).toMatchSnapshot()
    })

    it("should parse ex statements", () => {
        const result = parse([
            { str: "path", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "ex", type: "filterEx" },
            { str: "value", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: "", type: "eof" }
        ])

        expect(result).toMatchSnapshot()
    })

    it("should parse rgx statements", () => {
        const result = parse([
            { str: "path", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "rgx", type: "filterRgx" },
            { str: "value", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: "", type: "eof" }
        ])

        expect(result).toMatchSnapshot()
    })

    it("should parse lte statements", () => {
        const result = parse([
            { str: "path", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "lte", type: "filterLte" },
            { str: "value", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: "", type: "eof" }
        ])

        expect(result).toMatchSnapshot()
    })

    it("should throw an error when parsing invalid values", () => {
        expect(() =>
            parse([
                { str: "path", type: "path" },
                { str: "[", type: "filtersStart" },
                { str: "lte", type: "filterLte" },
                { str: "value", type: "error" },
                { str: "]", type: "filtersEnd" },
                { str: "", type: "eof" }
            ])
        ).toThrowErrorMatchingSnapshot()
    })

    it("should parse bundles in expressions correctly", () => {
        const result = parse([
            { str: "path", type: "path" },
            { str: "[", type: "filtersStart" },
            { str: "eq", type: "filterEq" },
            { str: "asd", type: "string" },
            { str: "|", type: "or" },
            { str: "gt", type: "filterGt" },
            { str: "asdf", type: "string" },
            { str: ",", type: "and" },
            { str: "lt", type: "filterLt" },
            { str: "asdf", type: "string" },
            { str: "]", type: "filtersEnd" },
            { str: "", type: "eof" }
        ])

        expect(result).toMatchSnapshot()
    })

    it("should throw an error when encountering eof at the end of an expression", () => {
        expect(() =>
            parse([
                { str: "path", type: "path" },
                { str: "[", type: "filtersStart" },
                { str: "eq", type: "filterEq" },
                { str: "asd", type: "string" },
                { str: "|", type: "or" },
                { str: "gt", type: "filterGt" },
                { str: "asdf", type: "string" },
                { str: "", type: "eof" }
            ])
        ).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error when encountering error after a filter", () => {
        expect(() =>
            parse([
                { str: "path", type: "path" },
                { str: "[", type: "filtersStart" },
                { str: "eq", type: "filterEq" },
                { str: "asd", type: "string" },
                { str: "]", type: "filtersEnd" },
                { str: "", type: "error" }
            ])
        ).toThrowErrorMatchingSnapshot()
    })
})
