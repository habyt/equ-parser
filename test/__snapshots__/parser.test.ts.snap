// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should parse bundles correctly 1`] = `
Array [
  Object {
    "expressions": Array [
      Object {
        "expressionType": "eq",
        "type": "expressionOperand",
        "value": "asd",
        "valueType": "string",
      },
    ],
    "path": "path",
    "type": "operand",
  },
  Object {
    "expressions": Array [
      Object {
        "expressionType": "eq",
        "type": "expressionOperand",
        "value": "bdf",
        "valueType": "string",
      },
    ],
    "path": "path2",
    "type": "operand",
  },
  Object {
    "operator": "and",
    "type": "operator",
  },
]
`;

exports[`parser should parse bundles in expressions correctly 1`] = `
Array [
  Object {
    "expressions": Array [
      Object {
        "expressionType": "eq",
        "type": "expressionOperand",
        "value": "asd",
        "valueType": "string",
      },
      Object {
        "expressionType": "gt",
        "type": "expressionOperand",
        "value": "asdf",
        "valueType": "string",
      },
      Object {
        "operator": "or",
        "type": "expressionOperator",
      },
      Object {
        "expressionType": "lt",
        "type": "expressionOperand",
        "value": "asdf",
        "valueType": "string",
      },
      Object {
        "operator": "and",
        "type": "expressionOperator",
      },
    ],
    "path": "path",
    "type": "operand",
  },
]
`;

exports[`parser should parse gte statements 1`] = `
Array [
  Object {
    "expressions": Array [
      Object {
        "expressionType": "gte",
        "type": "expressionOperand",
        "value": "value",
        "valueType": "string",
      },
    ],
    "path": "path",
    "type": "operand",
  },
]
`;

exports[`parser should parse lte statements 1`] = `
Array [
  Object {
    "expressions": Array [
      Object {
        "expressionType": "lte",
        "type": "expressionOperand",
        "value": "value",
        "valueType": "string",
      },
    ],
    "path": "path",
    "type": "operand",
  },
]
`;

exports[`parser should parse the example 1`] = `
Array [
  Object {
    "expressions": Array [
      Object {
        "expressionType": "eq",
        "type": "expressionOperand",
        "value": "foo",
        "valueType": "string",
      },
      Object {
        "expressionType": "eq",
        "type": "expressionOperand",
        "value": "bar",
        "valueType": "string",
      },
      Object {
        "operator": "or",
        "type": "expressionOperator",
      },
    ],
    "path": "name.first",
    "type": "operand",
  },
  Object {
    "expressions": Array [
      Object {
        "expressionType": "ct",
        "type": "expressionOperand",
        "value": "foo",
        "valueType": "string",
      },
      Object {
        "expressionType": "ct",
        "type": "expressionOperand",
        "value": "bar",
        "valueType": "string",
      },
      Object {
        "operator": "and",
        "type": "expressionOperator",
      },
    ],
    "path": "email",
    "type": "operand",
  },
  Object {
    "operator": "or",
    "type": "operator",
  },
  Object {
    "expressions": Array [
      Object {
        "expressionType": "gt",
        "type": "expressionOperand",
        "value": 4.5,
        "valueType": "number",
      },
      Object {
        "expressionType": "lt",
        "type": "expressionOperand",
        "value": -10,
        "valueType": "number",
      },
      Object {
        "operator": "and",
        "type": "expressionOperator",
      },
      Object {
        "expressionType": "eq",
        "type": "expressionOperand",
        "value": 15,
        "valueType": "number",
      },
      Object {
        "operator": "or",
        "type": "expressionOperator",
      },
    ],
    "path": "age",
    "type": "operand",
  },
  Object {
    "operator": "and",
    "type": "operator",
  },
]
`;

exports[`parser should throw a parse error 1`] = `"expected to have more items"`;

exports[`parser should throw a parse error for parsing expressions without a filter 1`] = `"unexpected {\\"str\\":\\"test\\",\\"type\\":\\"filtersEnd\\"}"`;

exports[`parser should throw a parse error on missing filter start 1`] = `"unexpected {\\"str\\":\\"bla\\",\\"type\\":\\"error\\"}"`;

exports[`parser should throw a parse error on unimplemented items 1`] = `"unexpected {\\"str\\":\\"test\\",\\"type\\":\\"error\\"}"`;

exports[`parser should throw an error when encountering eof at the end of an expression 1`] = `"unexpected {\\"str\\":\\"\\",\\"type\\":\\"eof\\"}"`;

exports[`parser should throw an error when encountering error after a filter 1`] = `"unexpected {\\"str\\":\\"\\",\\"type\\":\\"error\\"}"`;

exports[`parser should throw an error when parsing invalid values 1`] = `"unexpected {\\"str\\":\\"value\\",\\"type\\":\\"error\\"}"`;
